From eed36ee44db18c4d32896bbbfc390a2b9c4c715e Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Fri, 25 Jan 2019 12:34:16 +0530
Subject: [PATCH 0139/1852] ASoC: xlnx: enable multi sample rate support for
 I2S sound card

commit f188435b05d57456a10f7b90896b8e83227b4dc4 from
https://github.com/Xilinx/linux-xlnx.git

Patch enables I2S sound card with multiple sampling rates for playback
and capture. Values were based on the PMOD used in audio pipeline.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_pl_snd_card.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/sound/soc/xilinx/xlnx_pl_snd_card.c b/sound/soc/xilinx/xlnx_pl_snd_card.c
index 3f6b5bc6bf0d..5ba37dcbde1e 100644
--- a/sound/soc/xilinx/xlnx_pl_snd_card.c
+++ b/sound/soc/xilinx/xlnx_pl_snd_card.c
@@ -159,10 +159,13 @@ static int xlnx_i2s_card_hw_params(struct snd_pcm_substream *substream,
 		}
 	}
 
+	prv->mclk_val = prv->mclk_ratio * sample_rate;
 	clk_div = DIV_ROUND_UP(prv->mclk_ratio, 2 * ch * data_width);
 	ret = snd_soc_dai_set_clkdiv(cpu_dai, 0, clk_div);
+	if (ret)
+		return ret;
 
-	return ret;
+	return clk_set_rate(prv->mclk, prv->mclk_val);
 }
 
 static const struct snd_soc_ops xlnx_sdi_card_ops = {
-- 
2.31.1

